### PR TITLE
Add many test cases. Fix code to pass all.

### PR DESCRIPTION
**Pull request by TomOnTime**
- Add new test cases from MarkDown docs found in the wild.
- Add new test cases related to edge-cases involving hyphens.
- Verify all expected results against GitLab 7.13.2-ee

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -12,18 +12,28 @@ import "unicode"
...
+		case unicode.IsLetter(r) || unicode.IsNumber(r):
 			anchorName = append(anchorName, unicode.ToLower(r))
+			lastWasDash = false
 		default:
-			futureDash = true
 		}
 	}
+
+	// Special-case the situation where the entire string is spaces and
+	// special chars.  The spec is unclear what to do in this case. It is
```
Comment (dmitshur) [inline main.go]: Can you please elaborate on what spec you are referring to? It is not clear to me. Thanks!

**Comment by TomOnTime** [comment]
I've updated the comment to be more accurate. The "spec" I was referring to isn't really a spec, just the docs from the gitlab documentation.

**Comment by dmitshur** [comment]
> I've updated the comment to be more accurate. The "spec" I was referring to isn't really a spec, just the docs from the gitlab documentation.

All right, thanks. I'll try to review this weekend.

**Comment by dmitshur** [comment]
Reviewed at https://github.com/russross/blackfriday/pull/202#issuecomment-147149392.

**Comment by TomOnTime** [comment]
How's that?

**Comment by dmitshur** [comment]
The code looks okay, I just have some minor review comments I can post later.

However, two major points first:
1. I see that you've added the GitHub function largely as a response to me mentioning that GitLab should be in a dedicated function.
   
   However, who is the intended audience for the GitHub function? I think most people, myself included, would just use the standard `Create` in most cases.
   
   The GitHub function would just complicate matters and make users think about what they want to use. It also doesn't perfectly simulate GitHub rules for this, so it'd need to work to fix that.
   
   I don't have plans to use it at this time, and I don't have resources to develop it further (fix the known issues) and maintain it. I think it's best to remove it since it is unused, as I understand.
2. About the GitLab function, I understand you @TomOnTime want to use it.
   
   I think in this case the best thing to do is to add it to another repository, for example `github.com/TomOnTime/gitlabanchor`. That way you can use it by customizing `blackfriday` to use `gitlabanchor.Create`.
   
   I'd rather not accept it here because I don't have plans to use it and I don't want to (nor have the resources to) maintain it.

Thanks for the proposed contribution, I appreciate it. But I think it's best to place the `CreateGitLab` functionality in a separate repo that would be a better home for it. As far as I understand, there is no technical reason why it would need to be in this repo, since `blackfriday` will be customizable with a sanitize-anchor-name func from any Go package. Is that right?

What are your thoughts on that @TomOnTime?

**Comment by dmitshur** [comment]
Any thoughts on my last reply here, @TomOnTime?

**Comment by TomOnTime** [comment]
Hi!  Sorry, I haven't had time to look into this lately.  I'll see if I can get to it this weekend.

**Comment by TomOnTime** [comment]
I agree.  I've moved CreateGitLab and CreateGitHub to https://github.com/TomOnTime/markdownutils

Now that CreateGitLab and CreateGitHub are gone, all that's left of this PR is some unit tests that you may or may not be interested in.  Feel free to reject this PR if they are irrelevant.

**Comment by dmitshur** [comment]
Thank you, I think the current few and simple tests are sufficient for the needs of `sanitized_anchor_name`. It's primarily because the algorithm is so simple, it doesn't really need tests, just some examples.
